### ISSUE TITLE
"Cannot Open when State is Connecting."

### ISSUE THREAD

**Issue by Knaackee** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#85]
I have noticed a unexpected behavior.

In my controller i do a simple  `_databaseContext.Users.FirstOrDefault();` which works as expected (returns my first user).

But if i put a breakpoint on this line and move the cursor 
over `Users` in order to look into the `DbSet` i receive a empty list. After that i resume the program 
and get a Exception "Cannot Open when State is Connecting."

```
System.InvalidOperationException: Cannot Open when State is Connecting.
   at MySql.Data.MySqlClient.MySqlConnection.<OpenAsync>d__11.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
```

Any ideas?

Great work and thanks in advance!

**Comment by caleblloyd** [comment]
Could you post the full stack trace?

**Comment by dknaack** [comment]
If i put the cursor over the Users i receive 

```
System.Threading.ThreadAbortException: Thread was being aborted.
   at Windows.Foundation.Diagnostics.IAsyncCausalityTracerStatics.TraceOperationCreation(CausalityTraceLevel traceLevel, CausalitySource source, Guid platformId, UInt64 operationId, String operationName, UInt64 relatedContext)
   at System.Threading.Tasks.AsyncCausalityTracer.TraceOperationCreation(CausalityTraceLevel traceLevel, Int32 taskId, String operationName, UInt64 relatedContext)
```

On resume i receive


```
System.InvalidOperationException: Cannot Open when State is Connecting.
   at MySql.Data.MySqlClient.MySqlConnection.<OpenAsync>d__11.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.EntityFrameworkCore.Storage.Internal.MySqlRelationalConnection.<DoOpenAsync>d__36.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.EntityFrameworkCore.Storage.Internal.MySqlRelationalConnection.<BeginTransactionAsync>d__29.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.EntityFrameworkCore.Query.Internal.MySqlAsyncQueryingEnumerable`1.MySqlAsyncEnumerator.<MoveNext>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Linq.AsyncEnumerableExtensions.AsyncSelectEnumerable`2.AsyncSelectEnumerator.<MoveNext>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncLinqOperatorProvider.<_FirstOrDefault>d__82`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.EntityFrameworkCore.Query.Internal.TaskResultAsyncEnumerable`1.Enumerator.<MoveNext>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncLinqOperatorProvider.SelectAsyncEnumerable`2.SelectAsyncEnumerator.<MoveNext>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncLinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.<MoveNext>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at MyApp.Api.v1.Controller.MyController.<AssignTo>d__4.MoveNext() in Z:\Users\dknaack\Development\MyApp\myapp.api\src\MyApp.Api\v1\Controller\MyController.cs:line 156
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeActionFilterAsync>d__28.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeAsync>d__18.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Builder.RouterMiddleware.<Invoke>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Swashbuckle.SwaggerUi.Application.SwaggerUiMiddleware.<Invoke>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Swashbuckle.SwaggerUi.Application.RedirectMiddleware.<Invoke>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Swashbuckle.Swagger.Application.SwaggerMiddleware.<Invoke>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Cors.Infrastructure.CorsMiddleware.<Invoke>d__7.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at IdentityModel.AspNetCore.ScopeValidation.ScopeValidationMiddleware.<Invoke>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware`1.<Invoke>d__18.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware`1.<Invoke>d__18.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at IdentityServer4.AccessTokenValidation.IdentityServerAuthenticationMiddleware.<Invoke>d__7.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Server.IISIntegration.IISMiddleware.<Invoke>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Hosting.Internal.RequestServicesContainerMiddleware.<Invoke>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Server.Kestrel.Internal.Http.Frame`1.<RequestProcessingAsync>d__2.MoveNext()
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request finished in 91615.4192ms 200 
```

**Comment by caleblloyd** [comment]
The MySQL connection has a timeout of 15 seconds by default.  I wonder if the connection is being initiated but paused before it completes, then timing out.Does it go away if you add `ConnectionTimeout=0` to your MySQL Connection string?

**Comment by dknaack** [comment]
This does not solve my problem :( Its strange because after i jump over the breakpoint and the next gets hit i can view inside the DbSet's without any problem.

**Comment by Mats391** [comment]
I got the same error on occasions recently, but in a different scenario. Might help to track down the cause.
I opened two separate instances of DbContext from the same DbContextOptions instance. This caused both to use same connection and throw this error.

**Comment by caleblloyd** [comment]
@dknaack and @Mats391, how are you defining your `DbContextOptions`?

Are you passing a `Connection` or are you passing a `ConnectionString`?

If you're passing a `Connection`, I can see where passing the same instance of the connection to different DB Contexts would cause a problem.  You should make a new Connection for every Db Context.

If you're passing a `ConnectionString`, the library should generate a new connection for every DB Context.  All calls that open or close the connection are serialized using a semaphore, so it shouldn't be possible to call Open on it twice at the same time.  If it is possible, this is a bug that needs to be investigated.

All of this code is in [MySqlRelationalConnection.cs](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/master/src/Pomelo.EntityFrameworkCore.MySql/Storage/Internal/MySqlRelationalConnection.cs) for reference

**Comment by dknaack** [comment]
I use a `ConnectionString`

```
protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
{
    base.OnConfiguring(optionsBuilder);
    optionsBuilder.UseMySql("..");
}
```

**Comment by Mats391** [comment]
@caleblloyd I used a optionsbuilder and then just `new DbContext(builder.Options)`. So it used same connection. Fixed it by using ContextFactory that I pass connectionstring to which creates new optionsbuilder. However I did not use injected DbContext there, it was part of a console app.

@dknaack maybe try something like this in your ConfigureServices
```
string sqlConnectionString = Configuration.GetConnectionString("DataAccessMySqlProvider");
services.AddDbContext<MyDbContext>(options => options.UseMySql(sqlConnectionString));
```

**Comment by dknaack** [comment]
I tried your solution but it does not solve my problem. Thanks anyway!

**Comment by Kagamine** [comment]
@caleblloyd Is this issue related to ADO.NET Layer?

**Comment by caleblloyd** [comment]
@Kagamine I don't think so; it's probably something in  [MySqlRelationalConnection.cs](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/master/src/Pomelo.EntityFrameworkCore.MySql/Storage/Internal/MySqlRelationalConnection.cs) 

@dknaack could you post your `Startup.cs` and DbContext files?  You can take out any model references, I really just need the configuration-related aspects.

**Comment by dknaack** [comment]
@caleblloyd  sure.

`Startup.cs`        

```
// set db context
services.AddDbContext<MyDatabaseContext>(ServiceLifetime.Scoped);
```

The `MyDatabaseContext.cs` looks like this. (**NOTE**: `IdentityDbContext `comes from Microsoft.AspNetCore.Identity.EntityFrameworkCore)

    public class MyDatabaseContext : IdentityDbContext<ApplicationUser, IdentityRole<long>, long>
    {
        // a few DbSets<>
        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            base.OnConfiguring(optionsBuilder);
            switch (Environment.GetEnvironmentVariable("ASPNETCORE_ENVIRONMENT"))
            {
                case "Production":
                    optionsBuilder.UseMySql(ConnectionString.Production);
                    break;
                case "Staging":
                    optionsBuilder.UseMySql(ConnectionString.Staging);
                    break;
                default:
                    optionsBuilder.UseMySql(ConnectionString.Development);
                    break;
            }
        }

        protected override void OnModelCreating(ModelBuilder builder)
        {
            base.OnModelCreating(builder);
            // override table names
            builder.Entity<ApplicationUser>().ToTable("User");
            builder.Entity<IdentityRole<long>>().ToTable("Role");
            builder.Entity<IdentityUserClaim<long>>().ToTable("UserClaim");
            builder.Entity<IdentityRoleClaim<long>>().ToTable("RoleClaim");
            builder.Entity<IdentityUserRole<long>>().ToTable("UserRole");
            builder.Entity<IdentityUserLogin<long>>().ToTable("UserLogin");
            builder.Entity<IdentityUserToken<long>>().ToTable("UserToken");
        }
    }

**Comment by Kagamine** [comment]
Does this issue still occur?

**Comment by irontoby** [comment]
Hello, I am receiving this error in an ASP.NET Core 1.1 app using `Pomelo.EntityFrameworkCore.MySql` version `1.1.1-prerelease-10020`. It seems like after a while I start receiving the error and need to restart the app pool before connections will work again.

I am using the MySQL `DbContext` in both some custom auth middleware and from my main app (i.e. a class called from the controller). I've only seen the error in the middleware, but that's probably because it always fails there before it gets to the controller method. In both cases I'm using a `DbContext` instance returned by DI.

Let me know if there's anything else I can post to assist with tracking this down, thanks.

**Comment by vineus** [comment]
Hey @irontoby : I got exactly the same issue recently on a production machine (this error on a DbContext in a custom middleware) but in ASP core 2.0 using Pomelo 2.0.1

Did you find a workaround for the problem ?

thanks!

**Comment by irontoby** [comment]
@vineus no, sorry, I was never able to track this down further. I believe we just ended up decreasing the app pool restart timeout.

**Comment by mguinness** [comment]
The error "Cannot Open when State is Connecting" is thrown by the underlying library in [MySqlConnection](https://github.com/mysql-net/MySqlConnector/blob/master/src/MySqlConnector/MySql.Data.MySqlClient/MySqlConnection.cs#L163).  Is anyone able to share a git project which exhibits the behavior so we can try to recreate and take a look at?

**Comment by vineus** [comment]
Well, I had the error only once. It triggered when the connection to the MySQL server was briefly lost (network issue). Solved it by restarting my app.

**Comment by alihankocak** [comment]
I had same problem. Transient DBContext solved my problem. 
`services.AddDbContext<..>((serviceProvider, options) =>
            {
                options.UseMySql(Configuration.GetConnectionString("DefaultConnection"));
            }, ServiceLifetime.Transient);`

**Comment by kinglionsoft** [comment]
Same problem in NET6.0. A simple query in action:
``` c#
var bill = await _db.AgentBills.FindAsync(billId);
```

**Comment by mguinness** [comment]
Please create a new issue as this one is from 5 years ago.  Until someone can provide a full code example it will be hard to repro.
